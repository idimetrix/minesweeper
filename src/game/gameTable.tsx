import { Result } from 'antd';
import { TableOutlined } from '@ant-design/icons';
import { Button } from '@mui/material';

import { GameClient } from '../common/gameClient';
import { useGameTableStyles } from './gameTableStyles';

interface Properties {
  gameMap: string[];
}

const onCellClick = (y: number, x: number) => GameClient.socket.send(`open ${x} ${y}`);

export function GameTable({ gameMap }: Properties) {
  const classes = useGameTableStyles();

  const renderMap = (items: string[]) => {
    return items.map((item: string, rowIndex: number) => {
      const squares = [...item];
      const row = squares.map((square: any, columnIndex: number) => {
        const key = `square-${rowIndex}-${columnIndex}`;
        const testId = `square-${rowIndex}-${columnIndex}`;

        if (square !== 'â–¡') {
          return (
            <Button
              variant="outlined"
              color={square === '*' ? 'error' : 'success'}
              onClick={() => onCellClick(rowIndex, columnIndex)}
              key={key}
              className={classes.cell}
              data-testid={testId}
            >
              <p className={classes.text}>{square}</p>
            </Button>
          );
        }
        return (
          <Button
            variant="outlined"
            color="primary"
            onClick={() => onCellClick(rowIndex, columnIndex)}
            key={key}
            className={classes.cell}
            data-testid={testId}
          />
        );
      });

      return (
        // eslint-disable-next-line react/no-array-index-key
        <div className={classes.row} key={`square-row-${rowIndex}`}>
          {row}
        </div>
      );
    });
  };

  if (gameMap.length === 0) {
    return <Result icon={<TableOutlined />} title="Press START and start playing Minesweeper!" />;
  }

  return <>{renderMap(gameMap)}</>;
}
