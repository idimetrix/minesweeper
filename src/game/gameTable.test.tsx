// eslint-disable-next-line import/no-unresolved
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { cleanup } from '@testing-library/react';

import { GameTable } from './gameTable';

describe('<GameTable />', () => {
  afterEach(() => {
    jest.clearAllMocks();
    cleanup();
  });

  it('renders correctly and to match snapshot', () => {
    const gameMap: string[] = [];
    const tree = renderer.create(<GameTable gameMap={gameMap} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render empty state correctly', () => {
    const gameMap: string[] = [];
    const wrapper = mount(<GameTable gameMap={gameMap} />);

    expect(wrapper.text().includes('Press START and start playing Minesweeper!')).toBe(true);
  });

  it('should find at least 1 square cell', () => {
    const gameMap: string[] = [
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
    ];
    const wrapper = mount(<GameTable gameMap={gameMap} />);
    const element = wrapper.find('square-1-1');

    expect(element).toBeTruthy();
  });
});
