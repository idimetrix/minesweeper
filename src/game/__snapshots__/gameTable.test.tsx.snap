// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameTable /> renders correctly and to match snapshot 1`] = `
<div
  className="ant-result ant-result-info"
>
  <div
    className="ant-result-icon"
  >
    <span
      aria-label="table"
      className="anticon anticon-table"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="table"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M928 160H96c-17.7 0-32 14.3-32 32v640c0 17.7 14.3 32 32 32h832c17.7 0 32-14.3 32-32V192c0-17.7-14.3-32-32-32zm-40 208H676V232h212v136zm0 224H676V432h212v160zM412 432h200v160H412V432zm200-64H412V232h200v136zm-476 64h212v160H136V432zm0-200h212v136H136V232zm0 424h212v136H136V656zm276 0h200v136H412V656zm476 136H676V656h212v136z"
        />
      </svg>
    </span>
  </div>
  <div
    className="ant-result-title"
  >
    Press START and start playing Minesweeper!
  </div>
</div>
`;
